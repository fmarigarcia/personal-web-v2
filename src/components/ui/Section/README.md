# Section Component

A reusable wrapper component for page sections that ensures proper spacing and fills the viewport correctly accounting for the fixed navigation header.

## Features

- **Automatic navbar offset**: Accounts for the fixed header height using `SCROLL_OFFSET` constant
- **Full viewport height**: Option to make sections fill the entire viewport
- **Predefined backgrounds**: White, gray-50, or gradient backgrounds
- **Flexible content**: Accepts any React children
- **Responsive padding**: Consistent horizontal padding and responsive vertical spacing

## Props

```typescript
interface SectionProps {
    id: string; // Required: Section ID for navigation
    className?: string; // Optional: Additional CSS classes
    children: React.ReactNode; // Required: Section content
    fullHeight?: boolean; // Optional: Fill viewport height (default: false)
    backgroundColor?: 'white' | 'gray-50' | 'gradient'; // Optional: Background style (default: 'white')
}
```

## Usage Examples

### Basic Section (Content-based height)

```tsx
<Section id="about" backgroundColor="white">
    <div className="max-w-6xl mx-auto">
        <h2>About Us</h2>
        <p>Content here...</p>
    </div>
</Section>
```

### Full-height Section (Hero style)

```tsx
<Section
    id="hero"
    fullHeight
    backgroundColor="gradient"
    className="flex items-center justify-center"
>
    <div className="text-center">
        <h1>Welcome</h1>
        <button>Get Started</button>
    </div>
</Section>
```

### Gray Background Section

```tsx
<Section id="experience" backgroundColor="gray-50">
    <div className="max-w-6xl mx-auto">
        <h2>Experience</h2>
        {/* Timeline content */}
    </div>
</Section>
```

## Background Options

- **`white`**: Clean white background (`bg-white`)
- **`gray-50`**: Light gray background (`bg-gray-50`)
- **`gradient`**: Subtle gray gradient (`bg-gradient-to-br from-gray-50 to-gray-100`)

## Spacing Behavior

### Full Height Sections (`fullHeight={true}`)

- **Height**: `min-h-screen` - fills entire viewport
- **Top padding**: `${SCROLL_OFFSET}px` - accounts for fixed navbar
- **Bottom padding**: `0px` - content determines height
- **Content wrapper**: `min-h-screen flex flex-col justify-center py-20`

### Regular Sections (`fullHeight={false}`)

- **Height**: Content-driven (no min-height)
- **Top padding**: `0px` - no navbar offset needed
- **Bottom padding**: `5rem` - consistent section spacing
- **Content wrapper**: `py-20` - standard section padding

## Implementation Notes

1. **Navbar Integration**: Uses `SCROLL_OFFSET` constant (80px) from utils to account for fixed header
2. **Responsive Design**: Includes `px-6` for horizontal padding on all screen sizes
3. **Content Flow**: Regular sections stack naturally, full-height sections fill viewport
4. **TypeScript**: Fully typed with comprehensive interface

## Current Usage in Project

- **Hero**: `fullHeight + gradient` - Landing section that fills viewport
- **About**: `white` - Content section with standard padding
- **Experience**: `gray-50` - Alternating background for visual separation
- **Contact**: `gray-50` - Form section with consistent styling

This component centralizes section styling and ensures consistent spacing throughout the application while providing flexibility for different content types.
