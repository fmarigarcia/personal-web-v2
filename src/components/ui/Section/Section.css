/* Section component styles */
.section-navbar-offset {
    padding-top: 80px; /* Match SCROLL_OFFSET constant */
}

@media screen and (max-width: 768px) {
    .section-navbar-offset {
        padding-top: 0; /* Adjusted for smaller screens */
    }
}
