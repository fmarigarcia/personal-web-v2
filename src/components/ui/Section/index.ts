export { default } from './Section';
export { Section } from './Section';
