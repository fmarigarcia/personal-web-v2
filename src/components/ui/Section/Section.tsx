import React from 'react';
import clsx from 'clsx';
import './Section.css';

interface SectionProps {
    id: string;
    className?: string;
    children: React.ReactNode;
    fullHeight?: boolean;
    backgroundColor?: 'white' | 'gray-50' | 'gradient';
}

export const Section: React.FC<SectionProps> = ({
    id,
    className = '',
    children,
    fullHeight = false,
    backgroundColor = 'white',
}) => {
    // Background class mapping
    const backgroundClasses = {
        white: 'bg-white',
        'gray-50': 'bg-gray-50',
        gradient: 'bg-gradient-to-br from-gray-50 to-gray-100',
    };

    return (
        <section
            id={id}
            className={clsx(
                'relative px-6',
                {
                    'min-h-screen section-navbar-offset': fullHeight,
                    'pb-20': !fullHeight, // py-20 equivalent (5rem) when not full height
                },
                backgroundClasses[backgroundColor],
                className
            )}
        >
            {fullHeight ? (
                <div className="min-h-screen flex flex-col justify-center py-20">
                    {children}
                </div>
            ) : (
                <div className="py-20">{children}</div>
            )}
        </section>
    );
};

export default Section;
