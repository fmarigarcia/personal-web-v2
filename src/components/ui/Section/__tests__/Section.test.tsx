import { render, screen } from '@testing-library/react';
import { Section } from '../Section';

describe('Section', () => {
    const defaultProps = {
        id: 'test-section',
        children: <div>Test content</div>,
    };

    it('should render with basic props', () => {
        render(<Section {...defaultProps} />);

        const section = document.getElementById('test-section');
        expect(section).toBeInTheDocument();
        expect(section).toHaveAttribute('id', 'test-section');
        expect(screen.getByText('Test content')).toBeInTheDocument();
    });

    it('should apply white background by default', () => {
        render(<Section {...defaultProps} />);

        const section = document.getElementById('test-section');
        expect(section).toHaveClass('bg-white');
        expect(section).not.toHaveClass('min-h-screen');
    });

    it('should apply full height when fullHeight is true', () => {
        render(<Section {...defaultProps} fullHeight />);

        const section = document.getElementById('test-section');
        expect(section).toHaveClass('min-h-screen');
    });

    it('should apply gray-50 background', () => {
        render(<Section {...defaultProps} backgroundColor="gray-50" />);

        const section = document.getElementById('test-section');
        expect(section).toHaveClass('bg-gray-50');
    });

    it('should apply gradient background', () => {
        render(<Section {...defaultProps} backgroundColor="gradient" />);

        const section = document.getElementById('test-section');
        expect(section).toHaveClass('bg-gradient-to-br');
        expect(section).toHaveClass('from-gray-50');
        expect(section).toHaveClass('to-gray-100');
    });

    it('should apply custom className', () => {
        render(<Section {...defaultProps} className="custom-class" />);

        const section = document.getElementById('test-section');
        expect(section).toHaveClass('custom-class');
    });

    it('should have correct padding styles for fullHeight sections', () => {
        render(<Section {...defaultProps} fullHeight />);

        const section = document.getElementById('test-section');
        expect(section).toHaveStyle({ paddingTop: '80px' });
    });

    it('should have correct padding styles for regular sections', () => {
        render(<Section {...defaultProps} />);

        const section = document.getElementById('test-section');
        expect(section).toHaveStyle({ paddingTop: '0' });
        expect(section).toHaveStyle({ paddingBottom: '5rem' });
    });

    it('should match snapshot', () => {
        const { container } = render(<Section {...defaultProps} />);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot with full height and gradient', () => {
        const { container } = render(
            <Section
                {...defaultProps}
                fullHeight
                backgroundColor="gradient"
                className="custom-hero"
            />
        );
        expect(container.firstChild).toMatchSnapshot();
    });
});
