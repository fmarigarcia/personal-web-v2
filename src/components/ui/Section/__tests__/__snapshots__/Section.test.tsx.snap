// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section should match snapshot 1`] = `
<section
  class="relative px-6 min-h-screen section-navbar-offset bg-white"
  id="test-section"
>
  <div
    class="flex flex-col justify-center py-20"
  >
    <div>
      Test content
    </div>
  </div>
</section>
`;

exports[`Section should match snapshot with gradient 1`] = `
<section
  class="relative px-6 min-h-screen section-navbar-offset bg-gradient-to-br from-gray-50 to-gray-100 custom-hero"
  id="test-section"
>
  <div
    class="flex flex-col justify-center py-20"
  >
    <div>
      Test content
    </div>
  </div>
</section>
`;
